src/_gettsim/shared.py

```diff
@@ -260,3 +263,42 @@ def remove_group_suffix(col):
         out = out.removesuffix(f"_{g}")
 
     return out
+
+
+Key: TypeVar = TypeVar("Key")
+Out: TypeVar = TypeVar("Out")
+
+
+def join_numpy(
+    foreign_key: np.ndarray[Key], primary_key: np.ndarray[Key], target: np.ndarray[Out]
+) -> np.ndarray[Out]:
+    """
+    Given a foreign key, find the corresponding primary key, and return the target at
+    the same index as the primary key.
+
+    Parameters
+    ----------
+    foreign_key : np.ndarray[Key]
+        The foreign keys.
+    primary_key : np.ndarray[Key]
+        The primary keys.
+    target : np.ndarray[Out]
+        The targets in the same order as the primary keys.
+
+    Returns
+    -------
+    np.ndarray[Out]
+        The joined array.
+    """
+    if len(np.unique(primary_key)) != len(primary_key):
+        keys, counts = np.unique(primary_key, return_counts=True)
+        duplicate_foreign_keys = keys[counts > 1]
```

Comment:
```suggestion
        duplicate_primary_keys = keys[counts > 1]
```
Or am I missing something?